### DISCUSSION TITLE
Processing files and folders in-place

### DISCUSSION THREAD

**Discussion by ghost** [hairyhenderson/gomplate#1400] [Q&A] [answered]
Hello,

We found gomplate as a great alternative for `sed` and `envsubst` for Kubernetes manifest preprocessing.

We also found that we may process files and folders in-place:
```
# Same file
gomplate --file test.yaml --out test.yaml

# Same folder
gomplate --input-dir test-folder --output-dir test-folder
```

And we have some questions related to this
1. It is safe to preprocess files/folders in place?
2. Is there a switch to not use `--out`/`--output-dir` and just process same file/folder?
    ```
    gomplate --in-place --file test.yaml
    gomplate --in-place --input-dir test-folder
    ```

**Comment by hairyhenderson** [comment, marked as answer]
Hi @air3ijai, thanks for the questions - I've converted this to a Discussion since it does not appear to be a bug report or feature request.

> 1. It is safe to preprocess files/folders in place?

In general, it should be safe - gomplate reads the contents of each template into memory before rendering it, so the output file won't be written until after the input is read. Obviously, if you're processing a file in-place, it will be modified - that may not be considered "safe".

> 2. Is there a switch to not use `--out`/`--output-dir` and just process same file/folder?

No. It's pretty rare to do this as generally folks don't want templates to be modified, so I don't think it makes sense to add this as an implicit capability. But specifying the same file/dir for in/out works, so you can continue to do that.

Let me know if you have any other questions about this - hope this helped!

**Comment by ghost** [comment]
@hairyhenderson, thank you for moving that - will consider this for future questions.

We have a lot of stuff like
```
# envsubst
cat k8s/base/secrets.yaml | envsubst > k8s/base/secrets.yaml.tmp && mv k8s/base/secrets.yaml.tmp k8s/base/secrets.yaml

# sed
sed -i 's|{{NAMESPACE}}|${{ env.NAMESPACE }}|' k8s/overlays/${{ env.ENVIRONMENT_NAME }}/kustomization.yaml
```

and a simple alternative like this, looks very good
```
gomplate --file k8s/base/secrets.yaml --out k8s/base/secrets.yaml
gomplate --file k8s/overlays/${{ env.ENVIRONMENT_NAME }}/kustomization.yaml --out k8s/overlays/${{ env.ENVIRONMENT_NAME }}/kustomization.yaml
```

In our case we don't need to store template separately and manifests should be updated in-place, right before apply.

`--in-place` switch just can be more clean, without necessity to repeat same file/folder again
```
gomplate --in-place --file k8s/base/secrets.yaml
gomplate --in-place --file k8s/overlays/${{ env.ENVIRONMENT_NAME }}/kustomization.yaml
```

But for our case a solution like this, will replace a tens of commands
```
gomplate --input-dir k8s --output-dir k8s
```

Maybe someone else will ask same approach and you will consider a kind of `--in-place` for future releases.

Thank you for your work, Dave!
